From 931457e44e623089f0cc319988beb63e822fdfc9 Mon Sep 17 00:00:00 2001
From: Konstantin Klubnichkin <kitsok@yandex-team.ru>
Date: Mon, 16 Aug 2021 10:42:40 +0000
Subject: [PATCH 2/3] Introduce Yandex routines

AI1: Set BMC_INIT and BMC_READY
Start WDT3, set it to 180s
Detect platform from FRU

%% original patch: 0016-Introduce-Yandex-routines.patch
---
 board/aspeed/ast-g5/Makefile        |   1 +
 board/aspeed/ast-g5/ast-g5-yandex.c | 246 ++++++++++++++++++++++++++++++++++++
 board/aspeed/ast-g5/ast-g5.c        |  12 ++
 3 files changed, 259 insertions(+)
 create mode 100644 board/aspeed/ast-g5/ast-g5-yandex.c

diff --git a/board/aspeed/ast-g5/Makefile b/board/aspeed/ast-g5/Makefile
index df4e639..9f0cedb 100644
--- a/board/aspeed/ast-g5/Makefile
+++ b/board/aspeed/ast-g5/Makefile
@@ -1,2 +1,3 @@
 obj-y += ast-g5.o
+obj-y += ast-g5-yandex.o
 obj-y += ast-g5-irq.o
diff --git a/board/aspeed/ast-g5/ast-g5-yandex.c b/board/aspeed/ast-g5/ast-g5-yandex.c
new file mode 100644
index 0000000..f478ae5
--- /dev/null
+++ b/board/aspeed/ast-g5/ast-g5-yandex.c
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2021 Yandex LLC
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version
+ * 2 of the License, or (at your option) any later version.
+ */
+
+#include <common.h>
+#include <asm/io.h>
+#include <asm/arch/regs-scu.h>
+#include <asm/arch/ast_scu.h>
+#include <asm/arch/aspeed.h>
+#include <i2c.h>
+#include <watchdog.h>
+
+#include "ast-g5.h"
+
+static u32 machine_id;
+
+struct _fru_addr {
+	uchar bus;
+	uchar addr;
+	uchar width;
+};
+
+const struct _fru_addr frus[] = {
+	{ 0x03, 0x54, 1 },
+	{ 0x07, 0x51, 2 }
+};
+
+#define MACHINE_AI1  0x01
+#define MACHINE_MY81 0x02
+#define MACHINE_MZ81 0x03
+#define MACHINE_MY62 0x04
+#define MACHINE_MZB2 0x05
+#define MACHINE_GD1  0x06
+#define MACHINE_GD2  0x07
+#define MACHINE_G1   0x08
+
+#define MACHINE_UNK 0xff
+
+struct _machine_id {
+	char * pattern;
+	char * model;
+	uchar fru_bus;
+	uchar fru_addr;
+	char * fru_type;
+	char * host_cpu;
+	uchar machine_id;
+};
+
+#define MACHINE_IDS_SIZE 11
+static struct _machine_id machine_ids[] = {
+	{ "-AI1-",			  "ai1",  7, 0x51, "24c08",  "intel", MACHINE_AI1  },
+	{ "T175-N41-Y3N",	  "my81", 3, 0x54, "24c256", "intel", MACHINE_MY81 },
+	{ "T175-N40-Y3N",	  "mz81", 3, 0x54, "24c256", "amd",   MACHINE_MZ81 },
+	{ "-GI1-",			  "my62", 3, 0x54, "24c256", "intel", MACHINE_MY62 },
+	{ "-GA1-",			  "mzb2", 3, 0x54, "24c256", "amd",   MACHINE_MZB2 },
+	{ "Y4J-GD1-TY25-3P0", "gd1",  3, 0x54, "24c256", "none",  MACHINE_GD1  },
+	{ "Y4J-GE1-TY25-3P1", "gd1",  3, 0x54, "24c256", "none",  MACHINE_GD1  },
+	{ "Y4J-GD1-TY25-3Q0", "gd2",  3, 0x54, "24c256", "none",  MACHINE_GD2  },
+	{ "Y4J-GE1-TY25-3Q1", "gd2",  3, 0x54, "24c256", "none",  MACHINE_GD2  },
+	{ "-G1-",			  "g1",   3, 0x54, "24c256", "none",  MACHINE_G1   },
+	{ NULL,				  NULL,   0, 0x00, NULL,	  NULL,   MACHINE_UNK  },
+};
+
+struct _fru_common_header {
+	uchar version;
+	uchar offset_internal;
+	uchar offset_chassis;
+	uchar offset_board;
+	uchar offset_product;
+	uchar offset_multi;
+	uchar pad;
+	uchar crc;
+};
+
+/*--------------------------------------------------------------------
+ * @fn ai1_set_bmc_init_gpio
+ * @brief Pull low BMC_INIT GPIO line (GPIOE1)
+ *		Pull high BMC_READY GPIO line (GPIOE2)
+ * Author Kitsok
+ *--------------------------------------------------------------------*/
+void ai1_set_bmc_gpios(void)
+{
+	printf("BMC_INIT  -> 0\n");
+	printf("BMC_READY -> 1\n");
+	/* First set data register, then direction */
+	*(volatile ulong *)(0x1E780020) &= 0xfffffffd;
+	*(volatile ulong *)(0x1E780020) |= 0x00000004;
+	*(volatile ulong *)(0x1E780024) |= 0x00000006;
+}
+
+#define WDT3_TIMEOUT 180
+/* Init WDT3, set timeout */
+void init_wdt3(u32 seconds)
+{
+	printf("WDT3 enabled, %i seconds\n", seconds);
+	writel((u32) 1000000 * seconds, AST_WDT3_BASE + 0x04);
+	writel(0x4755, AST_WDT3_BASE + 0x08);
+	writel(0x13, AST_WDT3_BASE + 0x0c);
+}
+
+void ast_g5_yandex_early_init(void)
+{
+	printf("Yandex early init routine\n");
+	init_wdt3(WDT3_TIMEOUT);
+}
+
+uchar bus, addr, width;
+void search_eeprom(void)
+{
+	int fru_num, rc;
+	/* Find EEPROM */
+	bus = addr = 0xff;
+	for (fru_num = 0; fru_num < (sizeof(frus) / sizeof(struct _fru_addr)) ; fru_num++) {
+		i2c_set_bus_num(frus[fru_num].bus);
+		rc = i2c_probe(frus[fru_num].addr);
+		printf("Yandex: I2C bus:0x%02x Addr:0x%02x rc=%i\n",
+			frus[fru_num].bus, frus[fru_num].addr, rc);
+		if (rc == 0) {
+			bus = frus[fru_num].bus;
+			addr = frus[fru_num].addr;
+			width = frus[fru_num].width;
+			break;
+		}
+	}
+}
+
+void fru_print_common_header(struct _fru_common_header *p)
+{
+	printf("Yandex: FRU: Version				 0x%02x\n", p->version);
+	printf("Yandex: FRU: Internal area offset	 0x%02x\n", p->offset_internal);
+	printf("Yandex: FRU: Chassis area offset	 0x%02x\n", p->offset_chassis);
+	printf("Yandex: FRU: Board area offset		 0x%02x\n", p->offset_board);
+	printf("Yandex: FRU: Product area offset	 0x%02x\n", p->offset_product);
+	printf("Yandex: FRU: MultiRecord area offset 0x%02x\n", p->offset_multi);
+	printf("Yandex: FRU: CRC					 0x%02x\n", p->crc);
+}
+
+bool write_env = false;
+/* Detect machine by reading FRU
+ * 1. Search for FRU on I2C buses;
+ * 2. Read FRU header */
+int detect_machine(void)
+{
+	int i, rc, len;
+	u16 offset;
+	struct _fru_common_header fru_common_header;
+	uchar buffer[64];
+
+	search_eeprom();
+	if (bus == 0xff || addr == 0xff) {
+		printf("Yandex: No FRU found\n");
+		return -1;
+	}
+
+	printf("Yandex: FRU found bus:0x%02x Addr:0x%02x\n",
+			bus, addr);
+
+	memset((char *)buffer, 0, sizeof(buffer));
+	snprintf((char *)buffer, sizeof(buffer), "%x", bus);
+	setenv("machine_fru_bus", (const char *)buffer);
+
+	memset((char *)buffer, 0, sizeof(buffer));
+	snprintf((char *)buffer, sizeof(buffer), "%x", addr);
+	setenv("machine_fru_addr", (const char *)buffer);
+
+	write_env = true;
+
+	i2c_set_bus_num(bus);
+	rc = i2c_read(addr, 0x00, width, (uchar *) &fru_common_header, sizeof(struct _fru_common_header));
+	if (rc != 0) { printf("Yandex: Unable to read fru rc=%i\n", rc); return -1; }
+
+	/* Check if it's really FRU */
+	if (fru_common_header.version != 0x01 || fru_common_header.offset_product > 32) {
+		printf("Yandex: Broken FRU found\n");
+		setenv("machine_model", "unknown");
+		setenv("machine_fru_type", "24c256");
+		setenv("machine_host_cpu", "none");
+		write_env = true;
+		return -1;
+	}
+
+	fru_print_common_header(&fru_common_header);
+
+	/* First read 4 bytes to calculate neccessary offset of Product Area */
+	offset = 3 + (fru_common_header.offset_product * 8); // Points to Manufacturer name type/length
+	rc = i2c_read(addr, offset, width, (uchar *)buffer, 1);
+	if (rc != 0) { printf("Yandex: Unable to read fru rc=%i\n", rc); return -1; }
+	printf("Yandex: Product Area Manufacturer Name length 0x%02x\n", buffer[0]);
+
+	offset += (1 + (buffer[0] & 0x3f)); // Now points to Product Name type/length
+	rc = i2c_read(addr, offset, width, (uchar *)buffer, 1);
+	if (rc != 0) { printf("Yandex: Unable to read fru rc=%i\n", rc); return -1; }
+	printf("Yandex: Product Area Product Name length 0x%02x\n", buffer[0]);
+
+	/* Read Product Name itself */
+	offset++;
+	len = buffer[0] & 0x3f;
+	memset(&buffer, 0, sizeof(buffer));
+	rc = i2c_read(addr, offset, width, (uchar *)buffer, len);
+	if (rc != 0) { printf("Yandex: Unable to read fru rc=%i\n", rc); return -1; }
+	printf("Yandex: Product Name: %s\n", buffer);
+
+	/* Now find machine_id struct */
+	for(i = 0; machine_ids[i].pattern != NULL; i++) {
+		if (strstr((const char *)buffer, machine_ids[i].pattern) == NULL)
+			continue;
+
+		printf("Yandex: Found machine %s\n", machine_ids[i].model);
+		machine_id = machine_ids[i].machine_id;
+
+		/* Set environment variables */
+		setenv("machine_model", machine_ids[i].model);
+		setenv("machine_fru_type", machine_ids[i].fru_type);
+		setenv("machine_host_cpu", machine_ids[i].host_cpu);
+		write_env = true;
+
+		break;
+	}
+
+	if (machine_id == MACHINE_UNK)
+		return -1;
+
+	return 0;
+}
+
+void ast_g5_yandex_late_init(void)
+{
+
+	printf("Yandex late init routine\n");
+
+	machine_id = MACHINE_UNK;
+	detect_machine();
+
+	if (write_env)
+		saveenv();
+
+	/* Do machine-specific jobs */
+	if (machine_id == MACHINE_AI1) {
+		ai1_set_bmc_gpios();
+	}
+}
diff --git a/board/aspeed/ast-g5/ast-g5.c b/board/aspeed/ast-g5/ast-g5.c
index e791d21..8baaac3 100644
--- a/board/aspeed/ast-g5/ast-g5.c
+++ b/board/aspeed/ast-g5/ast-g5.c
@@ -20,11 +20,16 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+extern void ast_g5_yandex_early_init(void);
+extern void ast_g5_yandex_late_init(void);
+
 int board_init(void)
 {
 	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
 	gd->flags = 0;
 
+	ast_g5_yandex_early_init();
+
 	return 0;
 }
 
@@ -172,5 +177,12 @@ int board_early_init_f(void)
 int board_late_init(void)
 {
 	ast_g5_intel_late_init();
+
+#if defined(CONFIG_SYS_I2C)
+	i2c_init(I2C_ADAP->speed, I2C_ADAP->slaveaddr);
+#else
+	i2c_init(CONFIG_SYS_I2C_SPEED, CONFIG_SYS_I2C_SLAVE);
+#endif
+	ast_g5_yandex_late_init();
 	return 0;
 }
-- 
2.7.4

